libopenage/util/misc_test.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2017-2017 the openage authors. See copying.md for legal info.
+
+#include "misc.h"
+
+#include "../testing/testing.h"
+
+
+namespace openage {
+namespace util {
+namespace tests {
+
+
+void array_conversion() {
+	{
+		static const std::vector<uint8_t> test_array8{
+			0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef,
+			0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7
+		};
+		// The same array but as uint64_t
+		// ASDF: Can we just assume that everything is little-endian?
+		static const std::vector<uint64_t> test_array64{
+			0xefcdab8967452301, 0xf7f6f5f4f3f2f1f0
+		};
+
+		array8_to_array64(test_array8.data(), 16) == test_array64 or TESTFAIL;
+		array64_to_array8(test_array64.data(), 2) == test_array8 or TESTFAIL;
+	}
+	{
+		static const std::vector<uint8_t> test_array8{
+			0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef,
+			0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0x00, 0x00
+		};
+		// The same array but as uint64_t
+		// ASDF: Can we just assume that everything is little-endian?
+		static const std::vector<uint64_t> test_array64{
+			0xefcdab8967452301, 0x0000f5f4f3f2f1f0
+		};
+
+		array64_to_array8(test_array64.data(), 2) == test_array8 or TESTFAIL;
+		// Now I'm converting 14 instead of 16 bytes, the remaining ones should be
+		// filled with zeroes
+		array8_to_array64(test_array8.data(), 14) == test_array64 or TESTFAIL;
+	}
+	{
+		static const std::vector<uint8_t> test_array8{
+			0x01, 0x23,
+			0x45, 0x67, 0x89, 0xab, 0xcd, 0xef, 0xf0, 0xf1,
+			0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7
+		};
+		// The same array but as uint64_t
+		// ASDF: Can we just assume that everything is little-endian?
+		static const std::vector<uint64_t> expected{
+			0xf1f0efcdab896745, 0x0000f7f6f5f4f3f2
+		};
+
+		// Now I'm trying to convert misaligned bytes
+		// ASDF: Or It's just nonsense?
+		array8_to_array64(test_array8.data() + 2, 14) == expected or TESTFAIL;
```

Comment:
The problem I found with memcpy is that I need to reinterpret types anyway, for example, in function `array8_to_uint64`:

```
memcpy(reinterpret_cast<uint8_t *>(&result), start, count);
```

So now I'm just copying bytes into the uint64_t number memory, so far everything it's the same in every machine, but if I understand correctly, when I read that number as a whole the interpreted result depends on the endianness of the machine.

I've been looking into stackoverflow ([1](http://stackoverflow.com/questions/9695720/how-do-i-convert-a-64bit-integer-to-a-char-array-and-back), [2](http://stackoverflow.com/questions/2350099/how-to-convert-an-int-to-a-little-endian-byte-array)) and I think the best way to do this independently of endianness is by shifting as I did today.

I'm doing all this stuff because Siphash works on 64bit numbers, so at least I need `array8_to_uint64`, I thought that the remaining three functions are handy :stuck_out_tongue: 

Thank you for spending your time helping me, I really appreciate it